server/controllers/finance/cash.js

```diff
@@ -0,0 +1,247 @@
+      'cash.description, cash.cashbox_id, cash.is_caution, cash.user_id ' +
+    'FROM cash JOIN project ON cash.project_id = project.id ' +
+    'WHERE cash.uuid = ?;';
+
+  var cashItemsRecordSql =
+    'SELECT cash_item.uuid, cash_item.amount, cash_item.invoice_uuid ' +
+    'FROM cash_item WHERE cash_item.cash_uuid = ?;';
+
+  var cashDiscardRecordSql =
+    'SELECT cash_uuid FROM cash_discard WHERE cash_uuid = ?;';
+
+  return db.exec(cashRecordSql, [uuid])
+  .then(function (rows) {
+
+    if (rows.length === 0) {
+      throw codes.ERR_NOT_FOUND;
+    }
+
+    // store the record for return
+    record = rows[0];
+
+    return db.exec(cashItemsRecordSql, [uuid]);
+  })
+  .then(function (rows) {
+
+    // bind the cash items to the "items" property and return
+    record.items = rows;
+
+    return db.exec(cashDiscardRecordSql, [uuid]);
+  })
+  .then(function (rows) {
+
+    // if a linked cash_discard record exists, it means that this cash record
+    // has been reversed and we'll report that using a 'canceled' flag.
+    record.canceled = rows.length > 0;
+
+    return record;
+  });
+}
+
+/**
+ * GET /cash
+ * TODO - Query Paramters:
+ *   start={date}
+ *   end={date}
+ *   debtor={uuid}
+ *   type={ 'cash' | 'caution' },
+ *   project={id}
+ *
+ * Lists the cash payments with optional filtering parameters.
+ * @returns payments An array of { uuid, reference, date } JSON objects
+ */
+exports.list = function list(req, res, next) {
+  'use strict';
+
+  // base query
+  var sql =
+    'SELECT cash.uuid, CONCAT(project.abbr, cash.reference) AS reference, ' +
+      'cash.date, cash.amount '  +
+    'FROM cash JOIN project ON cash.project_id = project.id;';
+
+  db.exec(sql)
+  .then(function (rows) {
+    res.status(200).json(rows);
+  })
+  .catch(next)
+  .done();
+};
+
+/**
+ * GET /cash/:uuid
+ *
+ * Get the details of a particular cash payment.  This endpoint will return a
+ * record in the following format:
+ * {
+ *   uuid : "..",
+ *   items : [ {}, {} ]    // items hitting invoices, if applicable
+ *   is_caution : true/false,
+ *   canceled: true/false, // indicates if a cash_discard record exists
+ *   ...
+ *  }
+ */
+exports.details = function details(req, res, next) {
+  'use strict';
+
+  var uuid = req.params.uuid;
+
+  lookupCashRecord(uuid, req.codes)
+  .then(function (record) {
+    res.status(200).json(record);
+  })
+  .catch(next)
+  .done();
+};
+
+
+/**
+ * POST /cash
+ * Creates a cash payment against several sales or a cautionary payment.  The
+ * API also supports future offline functionality by either accepting a UUID or
+ * generating it if it is not present.
+ */
+exports.create = function create(req, res, next) {
+  'use strict';
+
+  var data = req.body.payment;
+
+  // generate a UUID if it not provided.
+  // @TODO - should we validate that this is an _actual_ uuid, or should this
+  // be sufficient?
+  data.uuid = data.uuid || uuid();
```

Comment:
It depends.  If we ever plan to support offline (client-side) creation of cash payments, sales, patients, etc, we must provide the option for the client to generate their own IDs.  If the client must have an active connection to the server at all times, we can limit creation to the server.

Ultimately, the database doesn't care - a UUID will be unique no matter where it is created.  This code just allows us to avoid that making the offline/online decision until we've gotten further into development.